File: apps/web/scripts/security-headers.mjs
function header(response, name) {
  return (response.headers.get(name) || '').trim()
}

function compact(value) {
  return value.toLowerCase().replace(/\s+/g, '')
}

function isPublicUrl(rawUrl) {
  const { hostname } = new URL(rawUrl)
  return hostname !== 'localhost' && hostname !== '127.0.0.1' && hostname !== '::1'
}

export function securityHeaderFailures(response, { surface, url }) {
  const csp = header(response, 'content-security-policy').toLowerCase()
  const permissions = compact(header(response, 'permissions-policy'))
  const failures = []

  if (header(response, 'x-frame-options').toUpperCase() !== 'DENY') {
    failures.push('x-frame-options=DENY')
  }
  if (header(response, 'x-content-type-options').toLowerCase() !== 'nosniff') {
    failures.push('x-content-type-options=nosniff')
  }
  if (header(response, 'referrer-policy').toLowerCase() !== 'no-referrer') {
    failures.push('referrer-policy=no-referrer')
  }
  if (new URL(url).protocol === 'https:' && !/^max-age=\d+/i.test(header(response, 'strict-transport-security'))) {
    failures.push('strict-transport-security=max-age')
  }
  if (!csp.includes("frame-ancestors 'none'")) failures.push("csp frame-ancestors 'none'")
  if (!csp.includes("object-src 'none'")) failures.push("csp object-src 'none'")

  if (surface === 'api') {
    if (!csp.includes("default-src 'none'")) failures.push("csp default-src 'none'")
    if (!csp.includes("base-uri 'none'")) failures.push("csp base-uri 'none'")
    if (!csp.includes("form-action 'none'")) failures.push("csp form-action 'none'")
    for (const directive of ['camera=()', 'microphone=()', 'display-capture=()', 'geolocation=()', 'payment=()']) {
      if (!permissions.includes(directive)) failures.push(`permissions-policy ${directive}`)
    }
  } else if (surface === 'web') {
    if (!csp.includes("default-src 'self'")) failures.push("csp default-src 'self'")
    if (!csp.includes("script-src 'self' 'wasm-unsafe-eval'")) {
      failures.push("csp script-src 'wasm-unsafe-eval'")
    }
    if (!csp.includes("base-uri 'self'")) failures.push("csp base-uri 'self'")
    if (!csp.includes("form-action 'self'")) failures.push("csp form-action 'self'")
    for (const directive of [
      'camera=(self)',
      'microphone=(self)',
      'display-capture=(self)',
      'geolocation=()',
      'payment=()',
    ]) {
      if (!permissions.includes(directive)) failures.push(`permissions-policy ${directive}`)
    }
    if (isPublicUrl(url) && /(?:localhost|127\.0\.0\.1)/i.test(csp)) {
      failures.push('production CSP contains a loopback target')
    }
  } else {
    failures.push(`unknown security header surface: ${surface}`)
  }

  return failures
}

export function assertSecurityHeaders(response, options) {
  const failures = securityHeaderFailures(response, options)
  if (failures.length > 0) {
    throw new Error(`${options.surface} security headers missing/invalid: ${failures.join(', ')}`)
  }
}

File: apps/server/src/middleware/security_headers.rs
use axum::{
    http::{HeaderName, HeaderValue},
    response::Response,
};

const CONTENT_SECURITY_POLICY: &str =
    "default-src 'none'; base-uri 'none'; form-action 'none'; frame-ancestors 'none'; object-src 'none'";
const PERMISSIONS_POLICY: &str =
    "camera=(), microphone=(), display-capture=(), geolocation=(), payment=()";

pub async fn apply(mut response: Response) -> Response {
    let headers = response.headers_mut();
    let csp_header = HeaderName::from_static("content-security-policy");
    if !headers.contains_key(&csp_header) {
        headers.insert(
            csp_header,
            HeaderValue::from_static(CONTENT_SECURITY_POLICY),
        );
    }
    headers.insert(
        HeaderName::from_static("strict-transport-security"),
        HeaderValue::from_static("max-age=31536000"),
    );
    headers.insert(
        HeaderName::from_static("x-frame-options"),
        HeaderValue::from_static("DENY"),
    );
    headers.insert(
        HeaderName::from_static("x-content-type-options"),
        HeaderValue::from_static("nosniff"),
    );
    headers.insert(
        HeaderName::from_static("referrer-policy"),
        HeaderValue::from_static("no-referrer"),
    );
    headers.insert(
        HeaderName::from_static("permissions-policy"),
        HeaderValue::from_static(PERMISSIONS_POLICY),
    );
    response
}

#[cfg(test)]
mod tests {
    use axum::{body::Body, http::StatusCode, response::Response};

    use super::*;

    #[tokio::test]
    async fn applies_the_api_security_header_contract() {
        let response = apply(
            Response::builder()
                .status(StatusCode::BAD_REQUEST)
                .body(Body::empty())
                .unwrap(),
        )
        .await;
        let headers = response.headers();

        assert_eq!(headers["x-frame-options"], "DENY");
        assert_eq!(headers["x-content-type-options"], "nosniff");
        assert_eq!(headers["referrer-policy"], "no-referrer");
        assert!(headers["content-security-policy"]
            .to_str()
            .unwrap()
            .contains("frame-ancestors 'none'"));
        assert!(headers["permissions-policy"]
            .to_str()
            .unwrap()
            .contains("microphone=()"));
    }

    #[tokio::test]
    async fn preserves_a_stricter_route_specific_csp() {
        let mut response = Response::builder()
            .status(StatusCode::OK)
            .body(Body::empty())
            .unwrap();
        response.headers_mut().insert(
            HeaderName::from_static("content-security-policy"),
            HeaderValue::from_static(
                "default-src 'none'; script-src 'nonce-test'; style-src 'nonce-test'",
            ),
        );

        let response = apply(response).await;

        assert_eq!(
            response.headers()["content-security-policy"],
            "default-src 'none'; script-src 'nonce-test'; style-src 'nonce-test'"
        );
        assert_eq!(response.headers()["x-frame-options"], "DENY");
    }
}
